File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size() ;
        vector<int> pref(n+1, 1);
        vector<int> suff(n+1, 1);
        int p=1, s = 1;
        for (int i = 0; i < n; i++){
            int k = nums[i];
            pref[i] = k * p;
            p = pref[i];
            suff[n-1-i] = s * nums[n-1-i];
            s = s * nums[n-1-i];
        }
        vector<int> ans(n);
        for (int i = 0; i < n; i++){
            if (i == 0){
                ans[i] = 1 * suff[1];
            }else {
                ans[i] = pref[i-1] * suff[i+1];
            }
        }
        return ans;
        
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [0] * n
        x=1
        for i in range(n):
            res[i] = x
            x *= nums[i]
            
        y = 1
        for i in reversed(range(n)):
            res[i] *= y
            y *= nums[i]
            
        return res
        
        